internal/pkg/agent/application/upgrade/details/details.go

```diff
@@ -0,0 +1,103 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package details
+
+import "sync"
+
+// Observer is a function that will be called with upgrade details
+type Observer func(details *Details)
+
+// Details consists of details regarding an ongoing upgrade.
+type Details struct {
+	TargetVersion string          `yaml:"target_version" json:"target_version"`
+	State         State           `yaml:"state" json:"state"`
+	ActionID      string          `yaml:"action_id" json:"action_id"`
+	Metadata      DetailsMetadata `yaml:"metadata" json:"metadata"`
+
+	observers []Observer
+	mu        sync.RWMutex
+}
+
+// DetailsMetadata consists of metadata relating to a specific upgrade state
+type DetailsMetadata struct {
+	ScheduledAt     string  `yaml:"scheduled_at" json:"scheduled_at"`
+	DownloadPercent float64 `yaml:"download_percent" json:"download_percent"`
+	FailedState     State   `yaml:"failed_state" json:"failed_state"`
```

Comment:
The idea here is to capture the state the upgrade was in if/when it failed, because if/when an upgrade fails, we will set `details.State` to `UPG_FAILED`.

We also capture any error message associated with the failure, in `details.Metadata.ErrorMsg`. See also the `Fail()` convenience method that handles this work.

Reply:
I saw the `Fail` after writing this comment and the naming of this field does not help. 
If I understand correctly, this state is the state the upgrade was in just before shifting to the error/failure state.
If that's the case why not keeping the error in the struct and the state as it was before? (That can be mapped to a failure state for fleet or other consumers separately).
Or we could keep a history of the states the upgrade went to with a start/end time for each, maybe a state... 
What about having a global `Success/Failure` outcome and keep the last state with the error ?

The problem is that this DetailsMetadata is public and mapped to json (for Fleet I imagine) so any of the changes above would be a change in the interface and I understand if we need to keep it because it is already in the contract.

Can I suggest that we change the name to `LastState` or `LastEnteredState` or maybe add a comment to this explaining exactly what it's for ? (preserving the state prior to the error since we are going to force that to `UPG_FAILED`) ?

Reply:
I'm resistant to renaming this field or changing the structure of `UpgradeDetails` unless you feel very strongly about it.  It impacts the same structure in the Fleet Server and Kibana codebases as well.  This is why we had an opportunity to solidify this structure and field naming during the design phase (doc was circulated in mid-May with a TTL of almost 2 weeks).  We should've done any renaming/restructuring at that time.  But again, if you feel very strongly against the current structure, we can put implementation on hold, reopen discussion again with relevant stakeholders, and make the necessary changes in all the codebases.

If you don't feel strongly about the structure/names, I'm happy to add the explanatory comment.